File: node/src/inmemorycheckpointer.test.ts
/*
 * Copyright 2021 IBM All Rights Reserved.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { DefaultCheckPointers } from './defaultcheckpointers';

    describe('Inmemory checkpointer', () => {

        it('No inmemory checkpointer exist: expected to initialize checkpointer state ', async () => {

            const inMemoryCheckPointer = DefaultCheckPointers.inMemory();

            const actualBlockNumber = await inMemoryCheckPointer.getBlockNumber();
            expect(actualBlockNumber).toBeUndefined();

            const actualTxnIDS = await inMemoryCheckPointer.getTransactionIds();
            expect(actualTxnIDS.length).toEqual(0);
        });


        it('Checkpointing only a block number in a fresh checkpointer gives block number & no transactions', async () => {

            const expectedBlockNumber = BigInt(101);

            const inMemoryCheckPointer = DefaultCheckPointers.inMemory();
            await inMemoryCheckPointer.checkpoint(expectedBlockNumber);

            const actualBlockNumber = await inMemoryCheckPointer.getBlockNumber();
            expect(actualBlockNumber).toEqual(expectedBlockNumber);

            const actualTxnIDS = await inMemoryCheckPointer.getTransactionIds();
            expect(actualTxnIDS.length).toEqual(0);
        });

        it('Checkpointing same block number and new transaction in used checkpointer gives block number and expected transactions', async () => {
            const expectedBlockNumber = BigInt(101);
            const expectedTxnIDs = ['txn1'];

            const inMemoryCheckPointer = DefaultCheckPointers.inMemory();
            await inMemoryCheckPointer.checkpoint(expectedBlockNumber)
            const actualBlockNumber = await inMemoryCheckPointer.getBlockNumber();
            expect(actualBlockNumber).toEqual(expectedBlockNumber);

            await inMemoryCheckPointer.checkpoint(expectedBlockNumber,'txn1');

            const actualTxnIDS = await inMemoryCheckPointer.getTransactionIds();
            expect(actualTxnIDS).toStrictEqual(expectedTxnIDs);
        });

        it('Checkpointing block and transaction in a fresh checkpointer, gives block number and transaction', async () => {

            const expectedBlockNumber = BigInt(101);
            const expectedTxnIDs = ['txn1'];

            const inMemoryCheckPointer = DefaultCheckPointers.inMemory();
            await inMemoryCheckPointer.checkpoint(expectedBlockNumber,'txn1')

            const actualBlockNumber = await inMemoryCheckPointer.getBlockNumber();
            expect(actualBlockNumber).toEqual(expectedBlockNumber);

            const actualTxnIDS = await inMemoryCheckPointer.getTransactionIds();
            expect(actualTxnIDS).toStrictEqual(expectedTxnIDs);
        });

        it('Checkpointing only a new block in used checkpointer gives new block number and no transactions', async () => {

            const blockNumber1 = BigInt(101);

            const blockNumber2 = BigInt(102);

            const expectedTxnIDs: any[] = [];

            const inMemoryCheckPointer = DefaultCheckPointers.inMemory();
            await inMemoryCheckPointer.checkpoint(blockNumber1,'txn1');

            let actualBlockNumber = await inMemoryCheckPointer.getBlockNumber();
            expect(actualBlockNumber).toStrictEqual(blockNumber1);

            await inMemoryCheckPointer.checkpoint(blockNumber2);
            actualBlockNumber = await inMemoryCheckPointer.getBlockNumber();
            expect(actualBlockNumber).toStrictEqual(blockNumber2);
            const actualTxnIDS = await inMemoryCheckPointer.getTransactionIds();
            expect(actualTxnIDS).toStrictEqual(expectedTxnIDs);

        });

            it('Checkpointing new block and transaction in used checkpointer gives new block and only new transaction', async () => {

                const blockNumber1 = BigInt(101);
                const blockNumber2 = BigInt(102);

                const expectedTxnIDs: string[] = ['txn2','txn3'];

                const inMemoryCheckPointer = DefaultCheckPointers.inMemory();
                await inMemoryCheckPointer.checkpoint(blockNumber1,'txn1');

                let actualBlockNumber = await inMemoryCheckPointer.getBlockNumber();
                expect(actualBlockNumber).toStrictEqual(blockNumber1);

                await inMemoryCheckPointer.checkpoint(blockNumber2);
                actualBlockNumber = await inMemoryCheckPointer.getBlockNumber();

                await inMemoryCheckPointer.checkpoint(blockNumber2,'txn2');
                await inMemoryCheckPointer.checkpoint(blockNumber2,'txn3');

                expect(actualBlockNumber).toStrictEqual(blockNumber2);
                const actualTxnIDS = await inMemoryCheckPointer.getTransactionIds();
                expect(actualTxnIDS).toStrictEqual(expectedTxnIDs);

            });
    })

File: pkg/checkpoint/inmemorycheckpointer_test.go
/*
Copyright 2022 IBM All Rights Reserved.

SPDX-License-Identifier: Apache-2.0
*/

package checkpoint

import (
	"testing"

	"github.com/stretchr/testify/require"
)

func TestInMemoryCheckpointer(t *testing.T) {
	t.Run("Initializes default checkpointer state when no checkpointer already exist", func(t *testing.T) {

		checkPointerInstance :=  new(InMemoryCheckpointer)
		blockNumber := checkPointerInstance.BlockNumber()
		require.EqualValues(t, uint64(0), blockNumber)
		require.EqualValues(t, "", checkPointerInstance.TransactionID())

	});
	t.Run("Checkpointing only a block number in a fresh checkpointer gives block number & blank transaction Id", func(t *testing.T) {

		expectedBlockNumber := uint64(101)
		checkPointerInstance :=  new(InMemoryCheckpointer)
		checkPointerInstance.CheckpointBlockNumber(expectedBlockNumber)
		actualblockNumber := checkPointerInstance.BlockNumber()
		require.EqualValues(t, expectedBlockNumber + 1 , actualblockNumber)
		require.EqualValues(t, "", checkPointerInstance.TransactionID())
	});
	t.Run("Checkpointing only a transaction id in a fresh checkpointer gives valid transaction Id and no blocknumber ", func(t *testing.T) {


		checkPointerInstance :=  new(InMemoryCheckpointer)
		checkPointerInstance.CheckpointTransaction("txn1")
		blockNumber := checkPointerInstance.BlockNumber()
		require.EqualValues(t, uint64(0), blockNumber)
		require.EqualValues(t, "txn1", checkPointerInstance.TransactionID())
	});

	t.Run("Checkpointing zero block number in a fresh checkpointer gives valid block number 1", func(t *testing.T) {

		expectedBlockNumber := uint64(0)
		checkPointerInstance :=  new(InMemoryCheckpointer)
		checkPointerInstance.CheckpointBlockNumber(expectedBlockNumber)
		actualblockNumber := checkPointerInstance.BlockNumber()
		require.EqualValues(t, expectedBlockNumber + 1 , actualblockNumber)
		require.EqualValues(t, "", checkPointerInstance.TransactionID())
	});

	t.Run("Checkpointing  block number and new transaction in used checkpointer unsets the  block number and gives expected transaction", func(t *testing.T) {

		expectedBlockNumber := uint64(101)
		checkPointerInstance :=  new(InMemoryCheckpointer)
		checkPointerInstance.CheckpointBlockNumber(expectedBlockNumber)
		checkPointerInstance.CheckpointTransaction("txn1")
		actualblockNumber := checkPointerInstance.BlockNumber()
		require.EqualValues(t, uint64(0), actualblockNumber)
		require.EqualValues(t, "txn1", checkPointerInstance.TransactionID())

	});

	t.Run("Checkpointing only a new block in used checkpointer gives new block number and blank transaction", func(t *testing.T) {

		blockNumber1 := uint64(101)
		blockNumber2 := uint64(102)

		checkPointerInstance :=  new(InMemoryCheckpointer)
		checkPointerInstance.CheckpointBlockNumber(blockNumber1)
		checkPointerInstance.CheckpointBlockNumber(blockNumber2)
		actualblockNumber := checkPointerInstance.BlockNumber()
		require.EqualValues(t, blockNumber2 + 1, actualblockNumber)
		require.EqualValues(t, "", checkPointerInstance.TransactionID())
	});
}